File: intellij-lsp/src/com/github/gtache/lsp/utils/FileUtils.scala
package com.github.gtache.lsp.utils

import java.io.File
import java.net.{URI, URL}

import com.github.gtache.lsp.utils.ApplicationUtils.computableWriteAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.{Document, Editor}
import com.intellij.openapi.fileEditor.{FileDocumentManager, FileEditorManager, OpenFileDescriptor, TextEditor}
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.project.{Project, ProjectUtil}
import com.intellij.openapi.vfs.{LocalFileSystem, VirtualFile}
import com.intellij.psi.PsiFile
import org.eclipse.lsp4j.TextDocumentIdentifier

/**
  * Various file / uri related methods
  */
object FileUtils {
  val os: OS.Value = if (System.getProperty("os.name").toLowerCase.contains("win")) OS.WINDOWS else OS.UNIX
  val COLON_ENCODED: String = "%3A"
  val SPACE_ENCODED: String = "%20"
  val URI_FILE_BEGIN: String = "file:"
  val URI_VALID_FILE_BEGIN: String = URI_FILE_BEGIN + "///"
  val URI_PATH_SEP: Char = '/'

  val LSP_ROOT_DIR: String = "lsp/"
  val LSP_LOG_DIR: String = LSP_ROOT_DIR + "log/"
  val LSP_CONFIG_DIR: String = LSP_ROOT_DIR + "conf/"
  private val LOG: Logger = Logger.getInstance(this.getClass)

  def extFromPsiFile(psiFile: PsiFile): String = {
    psiFile.getVirtualFile.getExtension
  }

  def editorFromPsiFile(psiFile: PsiFile): Editor = {
    editorFromVirtualFile(psiFile.getVirtualFile, psiFile.getProject)
  }

  def editorFromUri(uri: String, project: Project): Editor = {
    editorFromVirtualFile(URIToVFS(uri), project)
  }

  def editorFromVirtualFile(file: VirtualFile, project: Project): Editor = {
    FileEditorManager.getInstance(project).getAllEditors(file).collectFirst { case t: TextEditor => t.getEditor }.orNull
  }

  def openClosedEditor(uri: String, project: Project): (VirtualFile, Editor) = {
    val file = LocalFileSystem.getInstance().findFileByIoFile(new File(new URI(FileUtils.sanitizeURI(uri))))
    val fileEditorManager = FileEditorManager.getInstance(project)
    val descriptor = new OpenFileDescriptor(project, file)
    val editor: Editor = computableWriteAction(() => {
      fileEditorManager.openTextEditor(descriptor, false)
    })
    (file, editor)
  }

  /**
    * Returns a file type given an editor
    *
    * @param editor The editor
    * @return The FileType
    */
  def fileTypeFromEditor(editor: Editor): FileType = {
    FileDocumentManager.getInstance().getFile(editor.getDocument).getFileType
  }

  def extFromEditor(editor: Editor): String = {
    FileDocumentManager.getInstance().getFile(editor.getDocument).getExtension
  }

  /**
    * Transforms an editor (Document) identifier to an LSP identifier
    *
    * @param editor The editor
    * @return The TextDocumentIdentifier
    */
  def editorToLSPIdentifier(editor: Editor): TextDocumentIdentifier = {
    new TextDocumentIdentifier(editorToURIString(editor))
  }

  /**
    * Returns the URI string corresponding to an Editor (Document)
    *
    * @param editor The Editor
    * @return The URI
    */
  def editorToURIString(editor: Editor): String = {
    sanitizeURI(VFSToURI(FileDocumentManager.getInstance().getFile(editor.getDocument)))
  }

  /**
    * Returns the URI string corresponding to a VirtualFileSystem file
    *
    * @param file The file
    * @return the URI
    */
  def VFSToURI(file: VirtualFile): String = {
    if (file != null) {
      try {
        sanitizeURI(new URL(file.getUrl.replace(" ", SPACE_ENCODED)).toURI.toString)
      } catch {
        case e: Exception =>
          LOG.warn(e)
          LOG.warn("Caused by " + file.getUrl)
          null
      }
    } else null
  }

  /**
    * Fixes common problems in uri, mainly related to Windows
    *
    * @param uri The uri to sanitize
    * @return The sanitized uri
    */
  def sanitizeURI(uri: String): String = {
    if (uri != null) {
      val reconstructed: StringBuilder = StringBuilder.newBuilder
      var uriCp = new String(uri).replace(" ", SPACE_ENCODED) //Don't trust servers
      if (!uri.startsWith(URI_FILE_BEGIN)) {
        LOG.warn("Malformed uri : " + uri)
        uri //Probably not an uri
      } else {
        uriCp = uriCp.drop(URI_FILE_BEGIN.length).dropWhile(c => c == URI_PATH_SEP)
        reconstructed.append(URI_VALID_FILE_BEGIN)
        if (os == OS.UNIX) {
          reconstructed.append(uriCp).toString()
        } else {
          reconstructed.append(uriCp.takeWhile(c => c != URI_PATH_SEP))
          val driveLetter = reconstructed.charAt(URI_VALID_FILE_BEGIN.length)
          if (driveLetter.isLower) {
            reconstructed.setCharAt(URI_VALID_FILE_BEGIN.length, driveLetter.toUpper)
          }
          if (reconstructed.endsWith(COLON_ENCODED)) {
            reconstructed.delete(reconstructed.length - 3, reconstructed.length)
          }
          if (!reconstructed.endsWith(":")) {
            reconstructed.append(":")
          }
          reconstructed.append(uriCp.dropWhile(c => c != URI_PATH_SEP)).toString()
        }

      }
    } else null
  }

  /**
    * Transforms an URI string into a VFS file
    *
    * @param uri The uri
    * @return The virtual file
    */
  def URIToVFS(uri: String): VirtualFile = {
    LocalFileSystem.getInstance().findFileByIoFile(new File(new URI(sanitizeURI(uri))))
  }

  /**
    * Returns the project base dir uri given an editor
    *
    * @param editor The editor
    * @return The project whose the editor belongs
    */
  def editorToProjectFolderUri(editor: Editor): String = {
    pathToUri(editorToProjectFolderPath(editor))
  }

  def editorToProjectFolderPath(editor: Editor): String = {
    val project = editor.getProject
    if (project != null && !project.isDefault) {
      VFSToPath(ProjectUtil.guessProjectDir(project))
    } else null
  }

  def VFSToPath(file: VirtualFile): String = {
    new File(file.getPath).getAbsolutePath
  }

  /**
    * Transforms a path into an URI string
    *
    * @param path The path
    * @return The uri
    */
  def pathToUri(path: String): String = {
    sanitizeURI(new File(path.replace(" ", SPACE_ENCODED)).toURI.toString)
  }

  def uriToPath(uri: String): String = {
    new File(new URI(uri)).getAbsolutePath
  }

  def projectToUri(project: Project): String = {
    pathToUri(new File(project.getBasePath).getAbsolutePath)
  }

  def documentToUri(document: Document): String = {
    sanitizeURI(VFSToURI(FileDocumentManager.getInstance().getFile(document)))
  }

  def getAllOpenedEditors(project: Project): Seq[Editor] = {
    ApplicationUtils.computableReadAction(() => FileEditorManager.getInstance(project).getAllEditors().filter(e => e.isInstanceOf[TextEditor]).map(e => e.asInstanceOf[TextEditor].getEditor).toSeq)
  }

  /**
    * Object representing the OS type (Windows or Unix)
    */
  object OS extends Enumeration {
    type OS = Value
    val WINDOWS, UNIX = Value
  }

}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/utils/FileUtils.kt
package com.github.gtache.lsp.utils

import com.github.gtache.lsp.utils.ApplicationUtils.computableWriteAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.fileEditor.TextEditor
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.util.Computable
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiFile
import org.eclipse.lsp4j.TextDocumentIdentifier
import java.io.File
import java.net.URI
import java.net.URL

/**
 * Various file / uri related methods
 */
object FileUtils {
    private val os = if (System.getProperty("os.name").lowercase().contains("win")) OS.WINDOWS else OS.UNIX

    /**
     * Object representing the OS type (Windows or Unix)
     */
    private enum class OS {
        WINDOWS, UNIX
    }

    const val SPACE_ENCODED: String = "%20"
    private const val COLON_ENCODED: String = "%3A"
    private const val URI_FILE_BEGIN: String = "file:"
    private const val URI_VALID_FILE_BEGIN: String = "$URI_FILE_BEGIN///"
    private const val URI_PATH_SEP: Char = '/'

    private const val LSP_ROOT_DIR: String = "lsp/"
    const val LSP_LOG_DIR: String = LSP_ROOT_DIR + "log/"
    const val LSP_CONFIG_DIR: String = LSP_ROOT_DIR + "conf/"

    private val logger: Logger = Logger.getInstance(FileUtils::class.java)

    /**
     * Returns the extension of a [psiFile]
     */
    fun extFromPsiFile(psiFile: PsiFile): String? {
        return psiFile.virtualFile.extension
    }

    /**
     * Returns the editor of a [psiFile]
     */
    fun editorFromPsiFile(psiFile: PsiFile): Editor? {
        return editorFromVirtualFile(psiFile.virtualFile, psiFile.project)
    }

    /**
     * Returns the editor given an [uri] and a [project]
     */
    fun editorFromUri(uri: String, project: Project): Editor? {
        return uriToVFS(uri)?.let { editorFromVirtualFile(it, project) }
    }

    /**
     * Returns the editor given a virtual [file] and a [project]
     */
    fun editorFromVirtualFile(file: VirtualFile, project: Project): Editor? {
        return FileEditorManager.getInstance(project).getAllEditors(file).filterIsInstance<TextEditor>().map { t -> t.editor }.firstOrNull()
    }

    /**
     * Opens an editor given an [uri] and a [project] and returns a pair of (virtualFile, editor)
     */
    fun openClosedEditor(uri: String, project: Project): Pair<VirtualFile, Editor>? {
        val file = LocalFileSystem.getInstance().findFileByIoFile(File(URI(sanitizeURI(uri))))
        return if (file != null) {
            val descriptor = OpenFileDescriptor(project, file)
            val fileEditorManager = FileEditorManager.getInstance(project)
            val editor = computableWriteAction(Computable {
                fileEditorManager.openTextEditor(descriptor, false)
            })
            if (editor != null) {
                Pair(file, editor)
            } else null
        } else null
    }

    /**
     * Returns a file type given an [editor]
     */
    fun fileTypeFromEditor(editor: Editor): FileType? {
        return FileDocumentManager.getInstance().getFile(editor.document)?.fileType
    }

    /**
     * Returns the extension of a file opened in an [editor]
     */
    fun extFromEditor(editor: Editor): String? {
        return FileDocumentManager.getInstance().getFile(editor.document)?.extension
    }

    /**
     * Transforms an [editor] identifier to an LSP identifier
     */
    fun editorToLSPIdentifier(editor: Editor): TextDocumentIdentifier {
        return TextDocumentIdentifier(editorToURIString(editor))
    }

    /**
     * Returns the URI string corresponding to an [editor]
     */
    fun editorToURIString(editor: Editor): String? {
        val file = FileDocumentManager.getInstance().getFile(editor.document)
        return file?.let {
            vfsToURI(it)?.let { uri ->
                sanitizeURI(uri)
            }
        }
    }

    /**
     * Returns the URI string corresponding to a virtual [file]
     */
    fun vfsToURI(file: VirtualFile): String? {
        return if (!file.url.startsWith("mock") && !file.url.startsWith("dbSrc")) {
            try {
                sanitizeURI(URL(file.url.replace(" ", SPACE_ENCODED)).toURI().toString())
            } catch (e: Exception) {
                logger.warn(e)
                logger.warn("Caused by " + file.url)
                null
            }
        } else null
    }

    /**
     * Returns a sanitized uri by fixing common problems in a [uri], mainly related to Windows
     */
    fun sanitizeURI(uri: String): String {
        val reconstructed = StringBuilder()
        var uriCp = uri.replace(" ", SPACE_ENCODED) //Don't trust servers
        if (!uri.startsWith(URI_FILE_BEGIN)) {
            logger.warn("Malformed uri : $uri")
            return uri //Probably not an uri
        } else {
            uriCp = uriCp.drop(URI_FILE_BEGIN.length).dropWhile { c -> c == URI_PATH_SEP }
            reconstructed.append(URI_VALID_FILE_BEGIN)
            if (os == OS.UNIX) {
                return reconstructed.append(uriCp).toString()
            } else {
                reconstructed.append(uriCp.takeWhile { c -> c != URI_PATH_SEP })
                val driveLetter = reconstructed[URI_VALID_FILE_BEGIN.length]
                if (driveLetter.isLowerCase()) {
                    reconstructed.setCharAt(URI_VALID_FILE_BEGIN.length, driveLetter.uppercaseChar())
                }
                if (reconstructed.endsWith(COLON_ENCODED)) {
                    reconstructed.delete(reconstructed.length - 3, reconstructed.length)
                }
                if (!reconstructed.endsWith(":")) {
                    reconstructed.append(":")
                }
                return reconstructed.append(uriCp.dropWhile { c -> c != URI_PATH_SEP }).toString()
            }

        }
    }

    /**
     * Transforms an [uri] string into a VFS file
     */
    fun uriToVFS(uri: String): VirtualFile? {
        return LocalFileSystem.getInstance().findFileByIoFile(File(URI(sanitizeURI(uri))))
    }

    /**
     * Returns the project base dir uri given an [editor]
     */
    fun editorToProjectFolderUri(editor: Editor): String? {
        return editorToProjectFolderPath(editor)?.let {
            pathToUri(it)
        }
    }

    /**
     * Returns the project base dir path given an [editor]
     */
    fun editorToProjectFolderPath(editor: Editor): String? {
        val project = editor.project
        return if (project != null && !project.isDefault) {
            project.guessProjectDir()?.let {
                vfsToPath(it)
            }
        } else null
    }

    /**
     * Transform a virtual [file] to a path
     */
    fun vfsToPath(file: VirtualFile): String {
        return File(file.path).absolutePath
    }

    /**
     * Transforms a [path] into an URI string
     */
    fun pathToUri(path: String): String {
        return sanitizeURI(File(path.replace(" ", SPACE_ENCODED)).toURI().toString())
    }

    /**
     * Transforms an [uri] string to a path
     */
    fun uriToPath(uri: String): String {
        return File(URI(uri)).absolutePath
    }

    /**
     * Gets the uri of a [project]
     */
    fun projectToUri(project: Project): String? {
        return project.basePath?.let {
            pathToUri(File(it).absolutePath)
        }
    }

    /**
     * Gets the uri of a [document]
     */
    fun documentToUri(document: Document): String? {
        val file = FileDocumentManager.getInstance().getFile(document)
        return if (file != null) {
            vfsToURI(file)?.let {
                sanitizeURI(it)
            }
        } else null
    }

    /**
     * Retrieves all the currently opened editors for a [project]
     */
    fun getAllOpenedEditors(project: Project): List<Editor> {
        return ApplicationUtils.computableReadAction { FileEditorManager.getInstance(project).allEditors.filterIsInstance<TextEditor>().map { it.editor } }
    }


}